Integration tests fail with Mongo 3.6

<a href="https://github.com/ehlertjd"><img src="https://avatars2.githubusercontent.com/u/1509690?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [ehlertjd](https://github.com/ehlertjd)**
_Thursday Dec 14, 2017 at 15:04 GMT_
_Originally opened as https://github.com/scitran/core/issues/1037_

----

Steps to reproduce:
1. Change the `MONGO_VERSION` variable in `tests/bin/run-tests-docker.sh` to 3.6
1. Run integration tests

Expected results:
* Integration tests should pass

Actual results:
* There are 8 failures (of similar nature, relating to aggregation)

Test Failure:
```
__________________________________________________________________________________________________ test_collections __________________________________________________________________________________________________

data_builder = <conftest.DataBuilder object at 0x7f5382d3f4d0>, as_admin = <conftest.BaseUrlSession object at 0x7f5382d58210>, as_user = <conftest.BaseUrlSession object at 0x7f538282a050>

    def test_collections(data_builder, as_admin, as_user):
        session = data_builder.create_session()
        acquisition = data_builder.create_acquisition()
    
        # create collection
        r = as_admin.post('/collections', json={
            'label': 'SciTran/Testing'
        })
        assert r.ok
        collection = r.json()['_id']
    
        # get all collections w/ stats=true
        r = as_admin.get('/collections', params={'stats': 'true'})
>       assert r.ok
E       assert <Response [500]>.ok

tests/integration_tests/python/test_collection.py:14: AssertionError
```

Relevant Log snippet:
```
2017-12-13 22:42:37      scitran.api                  base.py    51:INFO Initialized request request_id=1b094d4c-1513204957
2017-12-13 22:42:37      scitran.api                  base.py   375:ERRO Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 570, in dispatch
    return method(*args, **kwargs)
  File "./api/handlers/collectionshandler.py", line 119, in get_all
    result = containerutil.get_stats(result, 'collections')
  File "./api/dao/containerutil.py", line 120, in get_stats
    result = config.db.command('aggregate', 'sessions', pipeline=pipeline).get('result', [])
  File "/usr/local/lib/python2.7/dist-packages/pymongo/database.py", line 481, in command
    codec_options, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/database.py", line 393, in _command
    allowable_errors)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/pool.py", line 211, in command
    read_concern)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/network.py", line 100, in command
    helpers._check_command_response(response_doc, msg, allowable_errors)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/helpers.py", line 196, in _check_command_response
    raise OperationFailure(msg % errmsg, code, response)
OperationFailure: command SON([('aggregate', 'sessions'), ('pipeline', [{'$match': {'_id': {'$in': []}}}, {'$project': {'non_compliant': {'$cond': [{'$eq': ['$satisfies_template', False]}, 1, 0]}, '_id': 1}}, {'$group': {'total': {'$sum': 1}, '_id': 1, 'noncompliant_count': {'$sum': '$non_compliant'}}}])]) on namespace scitran.$cmd failed: The 'cursor' option is required, except for aggregate with the explain argument
 request_id=1b094d4c-1513204957
127.0.0.1 - - [13/Dec/2017:22:42:37 +0000] "GET /api/collections?stats=true HTTP/1.1" 500 630 "-" "python-requests/2.18.4" request_id=1b094d4c-1513204957
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,5 +1,5 @@
 # base - install common dependencies & create directories
-FROM alpine:3.7 as base
+FROM alpine:3.8 as base
 ENV TERM=xterm
 RUN set -eux \
     && apk add --no-cache \

--- a/api/config.py
+++ b/api/config.py
@@ -260,20 +260,6 @@ def get_version():
 def get_item(outer, inner):
     return get_config()[outer][inner]
 
-def mongo_pipeline(table, pipeline):
-    """
-    Temporary philosophical dupe with reporthandler.py.
-    Execute a mongo pipeline, check status, return results.
-    A workaround for wonky pymongo aggregation behavior.
-    """
-
-    output = db.command('aggregate', table, pipeline=pipeline)
-    result = output.get('result')
-
-    if output.get('ok') != 1.0 or result is None:
-        raise Exception()
-
-    return result
 
 def get_auth(auth_type):
     return get_config()['auth'][auth_type]

--- a/api/dao/containerutil.py
+++ b/api/dao/containerutil.py
@@ -134,7 +134,7 @@ def get_stats(cont, cont_type):
         {'$group': {'_id': 1, 'noncompliant_count': {'$sum': '$non_compliant'}, 'total': {'$sum': 1}}}
     ]
 
-    result = config.db.command('aggregate', 'sessions', pipeline=pipeline).get('result', [])
+    result = list(config.db.sessions.aggregate(pipeline))
 
     if len(result) > 0:
         cont['session_count'] = result[0].get('total', 0)
@@ -154,7 +154,7 @@ def get_stats(cont, cont_type):
         {'$group': {'_id': 1, 'count': { '$sum': 1 }}}
     ]
 
-    result = config.db.command('aggregate', 'sessions', pipeline=pipeline).get('result', [])
+    result = list(config.db.sessions.aggregate(pipeline))
 
     if len(result) > 0:
         cont['subject_count'] = result[0].get('count', 0)

--- a/api/download.py
+++ b/api/download.py
@@ -451,17 +451,16 @@ class Download(base.RequestHandler):
             ]
 
             try:
-                result = config.db.command('aggregate', cont_name, pipeline=pipeline)
+                result = config.db[cont_name].aggregate(pipeline)
             except Exception as e: # pylint: disable=broad-except
                 self.log.warning(e)
                 self.abort(500, "Failure to load summary")
 
-            if result.get("ok"):
-                for doc in result.get("result"):
-                    type_ = doc['_id']
-                    if res.get(type_):
-                        res[type_]['count'] += doc.get('count',0)
-                        res[type_]['mb_total'] += doc.get('mb_total',0)
-                    else:
-                        res[type_] = doc
+            for doc in result:
+                type_ = doc['_id']
+                if res.get(type_):
+                    res[type_]['count'] += doc.get('count',0)
+                    res[type_]['mb_total'] += doc.get('mb_total',0)
+                else:
+                    res[type_] = doc
         return res

--- a/api/handlers/reporthandler.py
+++ b/api/handlers/reporthandler.py
@@ -114,31 +114,34 @@ class Report(object):
 
 
     @staticmethod
-    def _get_result_list(output):
+    def _get_result_list(cont_name, pipeline):
         """
         Helper function for extracting mongo aggregation results
 
-        Given the output of a mongo aggregation call, checks 'ok' field
-        If not 1.0, 'result' field does not exist or 'result' array is empty,
-        throws APIReportException
+        Runs a given mongo aggregation, throws APIReportException if
+        there was a mongo error or if there are no results.
         """
 
-        if output.get('ok') == 1.0:
-            result = output.get('result')
-            if result is not None and len(result) > 0:
-                return result
+        try:
+            result = list(config.db[cont_name].aggregate(pipeline))
+        except Exception as e:  # pylint: disable=broad-except
+            raise APIReportException(str(e))
 
-        raise APIReportException
+        if not result:
+            raise APIReportException('no results')
+
+        return result
+
 
     @staticmethod
-    def _get_result(output):
+    def _get_result(cont_name, pipeline):
         """
         Helper function for extracting a singular mongo aggregation result
 
         If more than one item is in the results array, throws APIReportException
         """
 
-        results = Report._get_result_list(output)
+        results = Report._get_result_list(cont_name, pipeline)
         if len(results) == 1:
             return results[0]
 
@@ -389,7 +392,7 @@ class ProjectReport(Report):
                 {'$group': {'_id': 1, 'count': { '$sum': 1 }}}
             ]
 
-            result = self._get_result(config.db.command('aggregate', 'sessions', pipeline=pipeline))
+            result = self._get_result('sessions', pipeline)
             project['subjects_count'] = result.get('count', 0)
 
             # Count subjects by sex
@@ -408,7 +411,7 @@ class ProjectReport(Report):
                                       'other':  {'$sum': '$other'}}}
             ]
             try:
-                result = self._get_result(config.db.command('aggregate', 'sessions', pipeline=pipeline))
+                result = self._get_result('sessions', pipeline)
             except APIReportException:
                 # Edge case when none of the subjects have a sex field
                 result = {}
@@ -429,7 +432,7 @@ class ProjectReport(Report):
                                                    'ethnicity': {'$last': '$subject.ethnicity'}}},
                 {'$group': {'_id': { 'sex': '$sex', 'race': '$race', 'ethnicity': '$ethnicity'}, 'count': {'$sum': 1}}}
             ]
-            results = self._get_result_list(config.db.command('aggregate', 'sessions', pipeline=pipeline))
+            results = self._get_result_list('sessions', pipeline)
 
             grid, total = self._process_demo_results(results, project['demographics_grid'])
             project['demographics_grid'] = grid
@@ -448,7 +451,7 @@ class ProjectReport(Report):
                 {'$group': {'_id': 1, 'over_18': {'$sum': '$over_18'}, 'under_18': {'$sum': '$under_18'}}}
             ]
             try:
-                result = self._get_result(config.db.command('aggregate', 'sessions', pipeline=pipeline))
+                result = self._get_result('sessions', pipeline)
             except APIReportException:
                 # Edge case when none of the subjects have an age field
                 result = {}
@@ -710,7 +713,7 @@ class UsageReport(Report):
         ]
 
         try:
-            results = self._get_result_list(config.db.command('aggregate', 'jobs', pipeline=pipeline))
+            results = self._get_result_list('jobs', pipeline)
         except APIReportException:
             results = []
 
@@ -733,7 +736,7 @@ class UsageReport(Report):
         ]
 
         try:
-            results = self._get_result_list(config.db.command('aggregate', 'sessions', pipeline=pipeline))
+            results = self._get_result_list('sessions', pipeline)
         except APIReportException:
             results = []
 
@@ -767,7 +770,7 @@ class UsageReport(Report):
             ]
 
             try:
-                results = self._get_result_list(config.db.command('aggregate', cont_name, pipeline=pipeline))
+                results = self._get_result_list(cont_name, pipeline)
             except APIReportException:
                 results = []
 
@@ -792,7 +795,7 @@ class UsageReport(Report):
             ]
 
             try:
-                results = self._get_result_list(config.db.command('aggregate', cont_name, pipeline=pipeline))
+                results = self._get_result_list(cont_name, pipeline)
             except APIReportException:
                 results = []
 
@@ -907,7 +910,7 @@ class UsageReport(Report):
                 ]
 
                 try:
-                    result = self._get_result(config.db.command('aggregate', cont_name, pipeline=pipeline))
+                    result = self._get_result(cont_name, pipeline)
                 except APIReportException:
                     result = None
 

--- a/api/jobs/queue.py
+++ b/api/jobs/queue.py
@@ -140,11 +140,14 @@ class Queue(object):
         log.info('respawned job %s as %s (attempt %d)', job.id_, new_id, new_job.attempt)
 
         # If job is part of batch job run, update batch jobs list
-        result = config.db.batch.update_one(
-            {'jobs': job.id_},
-            {'$pull': {'jobs': job.id_}, '$push': {'jobs': new_id}}
-        )
-        if result.modified_count == 1:
+        batch = config.db.batch.find_one({'jobs': job.id_})
+        if batch:
+            batch['jobs'].remove(job.id_)
+            batch['jobs'].append(new_id)
+            config.db.batch.update_one(
+                {'jobs': job.id_},
+                {'$set': {'jobs': batch['jobs']}}
+            )
             log.info('updated batch job list, replacing {} with {}'.format(job.id_, new_id))
 
         return new_id

--- a/bin/database.py
+++ b/bin/database.py
@@ -200,8 +200,8 @@ def upgrade_to_4():
         {'$group' : { '_id' : {'pid': '$project', 'code': '$subject.code'}, 'sids': {'$push': '$_id' }}}
     ]
 
-    subjects = config.db.command('aggregate', 'sessions', pipeline=pipeline)
-    for subject in subjects['result']:
+    subjects = config.db.sessions.aggregate(pipeline)
+    for subject in subjects:
 
         # Subjects without a code and sessions without a subject
         # will be returned grouped together, but all need unique IDs
@@ -553,8 +553,8 @@ def upgrade_to_17():
         {'$group' : { '_id' : {'pid': '$project', 'code': '$subject.code'}, 'sids': {'$push': '$_id' }}}
     ]
 
-    subjects = config.db.command('aggregate', 'sessions', pipeline=pipeline)
-    for subject in subjects['result']:
+    subjects = config.db.sessions.aggregate(pipeline)
+    for subject in subjects:
 
         # Subjects without a code and sessions without a subject
         # will be returned grouped together, but all need unique IDs
@@ -1396,8 +1396,8 @@ def upgrade_to_44():
         {'$group' : { '_id' : {'pid': '$project', 'code': '$subject.code'}, 'sids': {'$push': '$_id' }}}
     ]
 
-    subjects = config.db.command('aggregate', 'sessions', pipeline=pipeline)
-    for subject in subjects['result']:
+    subjects = config.db.sessions.aggregate(pipeline)
+    for subject in subjects:
 
         # Subjects without a code and sessions without a subject
         # will be returned grouped together, but all need unique IDs
@@ -1767,12 +1767,12 @@ def upgrade_to_53():
     Update rules to reference gears by id (`gear_id`) instead of name (`alg`)
     """
 
-    cursor = config.db.command('aggregate', 'gears', pipeline=[
+    cursor = config.db.gears.aggregate([
         {'$sort': {'gear.name': 1,
                    'created': -1}},
         {'$group': {'_id': '$gear.name',
                     'latest': {'$first': '$_id'}}}])
-    gear_name_to_id = {gear['_id']: str(gear['latest']) for gear in cursor['result']}
+    gear_name_to_id = {gear['_id']: str(gear['latest']) for gear in cursor}
 
     for rule in config.db.project_rules.find({'alg': {'$exists': True}}):
         config.db.project_rules.update_one(

--- a/bin/integrity_check.py
+++ b/bin/integrity_check.py
@@ -44,7 +44,7 @@ def session_length():
         {'$match': {'diff': {'$gt': 10800000}}}
     ]
 
-    results = config.db.command('aggregate', 'acquisitions', pipeline=pipeline)['result']
+    results = config.db.acquisition.aggregate(pipeline)
     if len(results) > 0:
         errors = True
         logging.warning('There are {} sessions that span 3 hours.'.format(len(results)))

--- a/bin/oneoffs/timezone_shift.py
+++ b/bin/oneoffs/timezone_shift.py
@@ -51,7 +51,7 @@ def shift_timezone(find, replace):
             {'$group': {'_id': '$session', 'min_timestamp': { '$min': '$timestamp' }}}
         ]
 
-        result = config.db.command('aggregate', 'acquisitions', pipeline=pipeline)['result'][0]
+        result = next(config.db.acquisition.aggregate(pipeline))
         min_timestamp = result.get('min_timestamp')
 
         # Update session to use earliest acquisition timestamp

--- a/requirements.txt
+++ b/requirements.txt
@@ -8,7 +8,7 @@ jsonschema==2.6.0
 Markdown==2.6.11
 prometheus_client==0.2.0
 psutil==5.4.6
-pymongo==3.6.1
+pymongo==3.7.1
 python-dateutil==2.7.3
 pytz==2018.4
 requests[security]==2.19.1

--- a/sdk/scripts/docker-tests.sh
+++ b/sdk/scripts/docker-tests.sh
@@ -83,7 +83,7 @@ main() {
         --network ${TEST_PREFIX} \
         --volume ${ROOT_DIR}/api:/var/scitran/code/api/api \
         --volume ${ROOT_DIR}/tests:/var/scitran/code/api/tests \
-        --env PRE_RUNAS_CMD='[ "$1" = uwsgi ] && mongod > /dev/null 2>&1 &' \
+        --env PRE_RUNAS_CMD='[ "$1" = uwsgi ] && mongod --bind_ip_all > /dev/null 2>&1 &' \
         --env SCITRAN_CORE_DRONE_SECRET=secret \
         --env SCITRAN_CORE_ACCESS_LOG_ENABLED=true \
         core:testing \

--- a/tests/bin/docker-tests.sh
+++ b/tests/bin/docker-tests.sh
@@ -67,7 +67,7 @@ main() {
         --name core-test-mongo \
         --network core-test \
         flywheel/core:testing \
-        mongod
+        mongod --bind_ip_all
 
     docker run --rm \
         --volume $(pwd):/pwd \

